readme.md
**Babbler** is a young [XMPP](http://xmpp.org) client library for Java SE 7 based on JAXB as XML processing technology.

It aims to provide good JavaDoc documentation, clean code, an easy to use API and a high level of software quality (which is currently ensured by 600+ unit tests).

You can find some more information [here](http://babbler-xmpp.blogspot.de/).

[HERE'S THE PROJECT SITE WITH DOCUMENTATION.](http://sco0ter.bitbucket.org/babbler/)

It supports most of the core specifications ([RFC 6120](http://xmpp.org/rfcs/rfc6120.html), [RFC 6121](http://xmpp.org/rfcs/rfc6121.html), [RFC 6122](http://xmpp.org/rfcs/rfc6122.html)).

Since this project is quite young, the API might change. Comments on the API are appreciated.

# Supported extensions
*(Only listing historical and standard tracks extensions with status 'Draft' or 'Final' that are applicable to XMPP clients)*

* ![alt supported][supported]           [XEP-0004: Data Forms](http://xmpp.org/extensions/xep-0004.html)
* ![alt supported][supported]           [XEP-0009: Jabber-RPC](http://xmpp.org/extensions/xep-0009.html)
* ![alt supported][supported]           [XEP-0012: Last Activity](http://xmpp.org/extensions/xep-0012.html)
* ![alt supported][supported]           [XEP-0013: Flexible Offline Message Retrieval](http://xmpp.org/extensions/xep-0013.html)
* ![alt supported][supported]           [XEP-0016: Privacy Lists](http://xmpp.org/extensions/xep-0016.html)
* ![alt supported][supported]           [XEP-0020: Feature Negotiation](http://xmpp.org/extensions/xep-0020.html)
* ![alt supported][supported]           [XEP-0030: Service Discovery](http://xmpp.org/extensions/xep-0030.html)
* ![alt supported][supported]           [XEP-0033: Extended Stanza Addressing](http://xmpp.org/extensions/xep-0033.html)
* ![alt not supported][not supported]   [XEP-0045: Multi-User Chat](http://xmpp.org/extensions/xep-0045.html)
* ![alt in development][in development] [XEP-0047: In-Band Bytestreams](http://xmpp.org/extensions/xep-0047.html)
* ![alt supported][supported]           [XEP-0048: Bookmarks](http://xmpp.org/extensions/xep-0048.html)
* ![alt supported][supported]           [XEP-0049: Private XML Storage](http://xmpp.org/extensions/xep-0049.html)
* ![alt not supported][not supported]   [XEP-0050: Ad-Hoc Commands](http://xmpp.org/extensions/xep-0050.html)
* ![alt supported][supported]           [XEP-0054: vcard-temp](http://xmpp.org/extensions/xep-0054.html)
* ![alt supported][supported]           [XEP-0055: Jabber Search](http://xmpp.org/extensions/xep-0055.html)
* ![alt in development][in development] [XEP-0059: Result Set Management](http://xmpp.org/extensions/xep-0059.html)
* ![alt in development][in development] [XEP-0060: Publish-Subscribe](http://xmpp.org/extensions/xep-0060.html)
* ![alt not supported][not supported]   [XEP-0065: SOCKS5 Bytestreams](http://xmpp.org/extensions/xep-0065.html)
* ![alt not supported][not supported]   [XEP-0066: Out of Band Data](http://xmpp.org/extensions/xep-0066.html)
* ![alt not supported][not supported]   [XEP-0070: Verifying HTTP Requests via XMPP](http://xmpp.org/extensions/xep-0070.html)
* ![alt not supported][not supported]   [XEP-0071: XHTML-IM](http://xmpp.org/extensions/xep-0071.html)
* ![alt not supported][not supported]   [XEP-0072: SOAP Over XMPP](http://xmpp.org/extensions/xep-0072.html)
* ![alt supported][supported]           [XEP-0077: In-Band Registration](http://xmpp.org/extensions/xep-0077.html)
* ![alt not supported][not supported]   [XEP-0079: Advanced Message Processing](http://xmpp.org/extensions/xep-0079.html)
* ![alt supported][supported]           [XEP-0080: User Location](http://xmpp.org/extensions/xep-0080.html)
* ![alt not supported][not supported]   [XEP-0084: User Avatar](http://xmpp.org/extensions/xep-0084.html)
* ![alt supported][supported]           [XEP-0085: Chat State Notifications](http://xmpp.org/extensions/xep-0085.html)
* ![alt supported][supported]           [XEP-0092: Software Version](http://xmpp.org/extensions/xep-0092.html)
* ![alt not supported][not supported]   [XEP-0095: Stream Initiation](http://xmpp.org/extensions/xep-0095.html)
* ![alt not supported][not supported]   [XEP-0096: SI File Transfer](http://xmpp.org/extensions/xep-0096.html)
* ![alt supported][supported]           [XEP-0106: JID Escaping](http://xmpp.org/extensions/xep-0106.html)
* ![alt supported][supported]           [XEP-0107: User Mood](http://xmpp.org/extensions/xep-0107.html)
* ![alt supported][supported]           [XEP-0108: User Activity](http://xmpp.org/extensions/xep-0108.html)
* ![alt supported][supported]           [XEP-0115: Entity Capabilities](http://xmpp.org/extensions/xep-0115.html)
* ![alt supported][supported]           [XEP-0118: User Tune](http://xmpp.org/extensions/xep-0118.html)
* ![alt supported][supported]           [XEP-0122: Data Forms Validation](http://xmpp.org/extensions/xep-0122.html)
* ![alt supported][supported]           [XEP-0124: Bidirectional-streams Over Synchronous HTTP (BOSH)](http://xmpp.org/extensions/xep-0124.html)
* ![alt supported][supported]           [XEP-0131: Stanza Headers and Internet Metadata](http://xmpp.org/extensions/xep-0131.html)
* ![alt not supported][not supported]   [XEP-0136: Message Archiving](http://xmpp.org/extensions/xep-0136.html)
* ![alt not supported][not supported]   [XEP-0137: Publishing Stream Initiation Requests](http://xmpp.org/extensions/xep-0137.html)
* ![alt supported][supported]           [XEP-0138: Stream Compression](http://xmpp.org/extensions/xep-0138.html)
* ![alt supported][supported]           [XEP-0141: Data Forms Layout](http://xmpp.org/extensions/xep-0141.html)
* ![alt in development][in development] [XEP-0144: Roster Item Exchange](http://xmpp.org/extensions/xep-0144.html)
* ![alt supported][supported]           [XEP-0145: Annotations](http://xmpp.org/extensions/xep-0145.html)
* ![alt supported][supported]           [XEP-0152: Reachability Addresses](http://xmpp.org/extensions/xep-0152.html)
* ![alt supported][supported]           [XEP-0153: vCard-Based Avatars](http://xmpp.org/extensions/xep-0153.html)
* ![alt not supported][not supported]   [XEP-0155: Stanza Session Negotiation](http://xmpp.org/extensions/xep-0155.html)
* ![alt supported][supported]           [XEP-0156: Discovering Alternative XMPP Connection Methods](http://xmpp.org/extensions/xep-0156.html)
* ![alt not supported][not supported]   [XEP-0158: CAPTCHA Forms](http://xmpp.org/extensions/xep-0158.html)
* ![alt not supported][not supported]   [XEP-0163: Personal Eventing Protocol](http://xmpp.org/extensions/xep-0163.html)
* ![alt not supported][not supported]   [XEP-0166: Jingle](http://xmpp.org/extensions/xep-0166.html)
* ![alt not supported][not supported]   [XEP-0167: Jingle RTP Sessions](http://xmpp.org/extensions/xep-0167.html)
* ![alt not supported][not supported]   [XEP-0171: Language Translation](http://xmpp.org/extensions/xep-0171.html)
* ![alt supported][supported]           [XEP-0172: User Nickname](http://xmpp.org/extensions/xep-0172.html)
* ![alt not supported][not supported]   [XEP-0174: Serverless Messaging](http://xmpp.org/extensions/xep-0174.html)
* ![alt not supported][not supported]   [XEP-0176: Jingle ICE-UDP Transport Method](http://xmpp.org/extensions/xep-0176.html)
* ![alt not supported][not supported]   [XEP-0177: Jingle Raw UDP Transport Method](http://xmpp.org/extensions/xep-0177.html)
* ![alt supported][supported]           [XEP-0184: Message Delivery Receipts](http://xmpp.org/extensions/xep-0184.html)
* ![alt in development][in development] [XEP-0191: Blocking Command](http://xmpp.org/extensions/xep-0191.html)
* ![alt not supported][not supported]   [XEP-0198: Stream Management](http://xmpp.org/extensions/xep-0198.html)
* ![alt supported][supported]           [XEP-0199: XMPP Ping](http://xmpp.org/extensions/xep-0199.html)
* ![alt supported][supported]           [XEP-0202: Entity Time](http://xmpp.org/extensions/xep-0202.html)
* ![alt supported][supported]           [XEP-0203: Delayed Delivery](http://xmpp.org/extensions/xep-0203.html)
* ![alt supported][supported]           [XEP-0206: XMPP Over BOSH](http://xmpp.org/extensions/xep-0206.html)
* ![alt supported][supported]           [XEP-0221: Data Forms Media Element](http://xmpp.org/extensions/xep-0221.html)
* ![alt supported][supported]           [XEP-0224: Attention](http://xmpp.org/extensions/xep-0224.html)
* ![alt not supported][not supported]   [XEP-0229: Stream Compression with LZW](http://xmpp.org/extensions/xep-0229.html)
* ![alt not supported][not supported]   [XEP-0231: Bits of Binary](http://xmpp.org/extensions/xep-0231.html)
* ![alt not supported][not supported]   [XEP-0249: Direct MUC Invitations](http://xmpp.org/extensions/xep-0249.html)
* ![alt supported][supported]           [XEP-0256: Last Activity in Presence](http://xmpp.org/extensions/xep-0256.html)
* ![alt not supported][not supported]   [XEP-0258: Security Labels in XMPP](http://xmpp.org/extensions/xep-0258.html)
* ![alt not supported][not supported]   [XEP-0260: Jingle SOCKS5 Bytestreams Transport Method](http://xmpp.org/extensions/xep-0260.html)
* ![alt not supported][not supported]   [XEP-0261: Jingle In-Band Bytestreams Transport Method](http://xmpp.org/extensions/xep-0261.html)
* ![alt not supported][not supported]   [XEP-0262: Use of ZRTP in Jingle RTP Sessions](http://xmpp.org/extensions/xep-0262.html)
* ![alt not supported][not supported]   [XEP-0266: Codecs for Jingle Audio](http://xmpp.org/extensions/xep-0266.html)
* ![alt not supported][not supported]   [XEP-0270: XMPP Compliance Suites 2010](http://xmpp.org/extensions/xep-0270.html)
* ![alt supported][supported]           [XEP-0297: Stanza Forwarding](http://xmpp.org/extensions/xep-0297.html)
* ![alt not supported][not supported]   [XEP-0301: In-Band Real Time Text](http://xmpp.org/extensions/xep-0301.html)
* ![alt supported][supported]           [XEP-0308: Last Message Correction](http://xmpp.org/extensions/xep-0308.html)

Supported experimental XEPs:

* ![alt supported][supported]           [XEP-0335: JSON Containers](http://xmpp.org/extensions/xep-0335.html)

Additionally following informational XEP documents are respected:

* ![alt supported][supported]           [XEP-0082: XMPP Date and Time Profiles](http://xmpp.org/extensions/xep-0082.html)
* ![alt supported][supported]           [XEP-0083: Nested Roster Groups](http://xmpp.org/extensions/xep-0083.html)
* ![alt supported][supported]           [XEP-0128: Service Discovery Extensions](http://xmpp.org/extensions/xep-0128.html)
* ![alt supported][supported]           [XEP-0170: Recommended Order of Stream Feature Negotiation](http://xmpp.org/extensions/xep-0170.html)
* ![alt supported][supported]           [XEP-0201: Best Practices for Message Threads](http://xmpp.org/extensions/xep-0201.html)



# License

This project is licensed under [MIT License](http://opensource.org/licenses/MIT).

# Getting started

## Creating a XMPP connection

There are two kinds of connection:

1. A [standard TCP connection](http://xmpp.org/rfcs/rfc6120.html#tcp)
2. A [BOSH connection](http://xmpp.org/extensions/xep-0124.html)

The following will create a normal connection to the given host:

```java
Connection connection = new TcpConnection("hostname", 5222);
```

If you want to establish a BOSH connection, use the following class instead:
```java
Connection connection = new BoshConnection("hostname", 5280);
```

In either case you get an abstract connection object, you can now work with.

## Preparing your connection

Before actually connecting to the server, you should setup your environment:

* Setting up event listeners in order to listen for incoming messages, roster and presence changes or to modify outgoing messages.
* Configure how features are negotiated, e.g. by setting up a `SSLContext`.
* Configure extensions, e.g.
 * Enable or disable certain extensions
 * Set an identity for the connection (Service Discovery)
 * ...


```java
// Setting a custom SSL context
connection.getSecurityManager().setSSLContext(sslContext);
// Listen for presence changes
connection.addPresenceListener(new PresenceListener() {
    @Override
    public void handle(PresenceEvent e) {
        if (e.isIncoming()) {
            // Handle incoming presence.
        }
    }
});
// Listen for messages
connection.addMessageListener(new MessageListener() {
    @Override
    public void handle(MessageEvent e) {
        // Handle outgoing or incoming message
    }
});
// Listen for roster pushes
connection.getRosterManager().addRosterListener(new RosterListener() {
    @Override
    public void rosterChanged(RosterEvent e) {
    }
});
```

## Connecting

If you want to connect to the server, you can do it like that:


```java
try {
   connection.connect();
} catch (IOException e) {
   // e.g. UnknownHostException
}
```

This will

* open the initial XMPP stream to the server.
* negotiate any features offered by the server, especially TLS.

## Authenticating and binding a resource

After connecting, you have to authenticate and bind a resource, in order to become a "connected resource". After that step you will be able to send message, presence and iq stanzas.


```java
try {
   connection.login("username", "password", "resource");
} catch (FailedLoginException e) {
   // Login failed, due to wrong username/password
}
```

## Establishing a presence session
After you are connected, authenticated and have bound a resource you should now establish a presence session, by sending [initial presence](http://xmpp.org/rfcs/rfc6121.html#presence-initial):


```java
connection.send(new Presence());
```

# Managing extensions
## Getting extensions from stanzas

```java
DelayedDelivery delayedDelivery = message.getExtension(DelayedDelivery.class);
```

```java
Attention attention = message.getExtension(Attention.class);
```
## Enabling and using extensions

```java
MessageDeliveryReceiptsManager messageDeliveryReceiptsManager = connection.getExtensionManager(MessageDeliveryReceiptsManager.class);
messageDeliveryReceiptsManager.setEnabled(true);
messageDeliveryReceiptsManager.addMessageDeliveredListener(new MessageDeliveredListener() {
    @Override
    public void messageDelivered(MessageDeliveredEvent e) {
        System.out.println("Message delivered: " + e.getMessageId());
    }
});
```

```java
LastActivityManager lastActivityManager = connection.getExtensionManager(LastActivityManager.class);
lastActivityManager.getLastActivity(Jid.valueOf("juliet@example.net"));
```


```java
SoftwareVersionManager softwareVersionManager = connection.getExtensionManager(SoftwareVersionManager.class);
SoftwareVersion softwareVersion = softwareVersionManager.getSoftwareVersion(Jid.valueOf("romeo@example.net"));
```

[supported]: /sco0ter/babbler/raw/tip/supported.png "Is supported"
[not supported]: /sco0ter/babbler/raw/tip/notsupported.png "Is not supported"
[in development]: /sco0ter/babbler/raw/tip/development.png "Is in development or planned"

changelog.md
# Changelog
---

## Version 0.2.0-SNAPSHOT

* RosterManager now supports (nested) contact groups
* The 'from' attribute of roster pushes are now checked
* Roster.Contact class refactored to class Contact (API change)	
* Added support for [XEP-0141: Data Forms Layout](http://xmpp.org/extensions/xep-0141.html) 
* Added support for [XEP-0122: Data Forms Validation](http://xmpp.org/extensions/xep-0122.html)
* Added support for [XEP-0335: JSON Containers](http://xmpp.org/extensions/xep-0335.html)
* Jid class improved (nodeprep, resourceprep, better escaping logic, caching for better performance, Comparable interface implemented)
* Added JID Escaping feature ("jid\\escaping") to feature list for Service Discovery.
* Restructured the project: It's now separated into modules:
** xmpp-core, which contains XML schema implementations and core classes. It could theoretically be useful for a server implementation, too.
** xmpp-client, which contains business logic, used by XMPP clients (e.g. connection logic, roster management, ...)
* Message, Presence and IQ classes have been moved from org.xmpp.stanza to org.xmpp.stanza.client package (API change).
* The base Connection class is now called XmppSession, which can have multiple connection methods. Each connection method is tried while connecting. That way a XMPP session can have a normal TcpConnection and an alternative BoshConnection, which is tried as fallback.
* Updated [XEP-0080: User Location](http://xmpp.org/extensions/xep-0080.html) from version 1.7 to 1.8.


## Version 0.1.0 (2014-03-22)

* Initial version

src/main/java/org/xmpp/Jid.java
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2014 Christian Schudt
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package org.xmpp;

import java.text.Normalizer;
import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * The implementation of the JID as described in <a href="http://xmpp.org/rfcs/rfc6122.html">Extensible Messaging and Presence Protocol (XMPP): Address Format</a>.
 * <p>
 * A JID consists of three parts:
 * <p>
 * [ localpart "@" ] domainpart [ "/" resourcepart ]
 * </p>
 * The easiest way to create a JID is to use the {@link #valueOf(String)} method:
 * <pre><code>
 * Jid jid = Jid.valueOf("juliet@capulet.lit/balcony");
 * </code></pre>
 * You can then get the parts from it via the respective methods:
 * <pre><code>
 * String local = jid.getLocal(); // juliet
 * String domain = jid.getDomain(); // capulet.lit
 * String resource = jid.getResource(); // balcony
 * </code></pre>
 * This class overrides <code>equals()</code> and <code>hashCode()</code>, so that different instances with the same value are equal:
 * <pre><code>
 * Jid.valueOf("romeo@capulet.lit/balcony").equals(Jid.valueOf("romeo@capulet.lit/balcony")); // true
 * </code></pre>
 * This class also supports <a href="http://xmpp.org/extensions/xep-0106.html">XEP-0106: JID Escaping</a>, i.e.
 * <pre><code>
 * Jid.valueOf("d'artagnan@musketeers.lit")
 * </code></pre>
 * is escaped as <code>d\\27artagnan@musketeers.lit</code>.
 *
 * @author Christian Schudt
 */
public final class Jid implements Comparable<Jid> {

    /**
     * Escapes all disallowed characters and also backslash, when followed by a defined hex code for escaping. See 4. Business Rules.
     */
    private static final Pattern ESCAPE_PATTERN = Pattern.compile("[ \"&'/:<>@]|\\\\(?=20|22|26|27|2f|3a|3c|3e|40|5c)");

    private static final Pattern UNESCAPE_PATTERN = Pattern.compile("\\\\(20|22|26|27|2f|3a|3c|3e|40|5c)");

    /**
     * Every character, which is not a letter, number, punctuation, symbol character, marker character or space.
     */
    private static final Pattern PROHIBITED_CHARACTERS = Pattern.compile("[^\\p{L}\\p{N}\\p{P}\\p{S}\\p{M}\\s]");

    private static final String DOMAIN_PART = "((?:(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\\-]*[a-zA-Z0-9])\\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\\-]*[A-Za-z0-9]))+)";

    private static final Pattern JID = Pattern.compile("^((.*?)@)?" + DOMAIN_PART + "(/(.*))?$");

    // B.1 Commonly mapped to nothing
    // Every space, except white-space (\u0020)
    private static final Pattern MAP_TO_NOTHING = Pattern.compile("([\u00AD\u034F\u1806\u180B\u180C\u180D\u200B\u200C\u200D\u2060\uFE00\uFE01\uFE0F\uFEFF])");

    private final String escapedLocal;

    private final String local;

    private final String domain;

    private final String resource;

    /**
     * Creates a bare JID with only the domain part.
     *
     * @param domain The domain.
     */
    public Jid(String domain) {
        this(null, domain, null);
    }

    /**
     * Creates a bare JID with only the local and domain part.
     *
     * @param local  The local part.
     * @param domain The domain part.
     */
    public Jid(String local, String domain) {
        this(local, domain, null);
    }

    /**
     * Creates a full JID with local, domain and resource part.
     *
     * @param local    The local part.
     * @param domain   The domain part.
     * @param resource The resource part.
     */
    public Jid(String local, String domain, String resource) {
        this(local, domain, resource, false, true);
    }

    private Jid(String local, String domain, String resource, boolean doUnescape, boolean prepareAndValidate) {
        String preparedNode;
        if (prepareAndValidate) {
            preparedNode = prepare(local, true);
            validateDomain(domain);
            validateLength(preparedNode, "local");
        } else {
            preparedNode = local;
        }
        String preparedResource = prepare(resource, false);
        validateLength(preparedResource, "resource");

        if (doUnescape) {
            this.local = unescape(preparedNode);
        } else {
            this.local = preparedNode;
        }
        this.escapedLocal = escape(this.local);
        this.domain = domain.toLowerCase();
        this.resource = preparedResource;
    }

    /**
     * Creates a JID from a string. The format must be
     * <blockquote><p>[ localpart "@" ] domainpart [ "/" resourcepart ]</p></blockquote>.
     * The input string will be escaped.
     *
     * @param jid The JID.
     * @return The JID.
     * @see <a href="http://xmpp.org/extensions/xep-0106.html">XEP-0106: JID Escaping</a>
     */
    public static Jid valueOf(String jid) {
        return valueOf(jid, false);
    }

    /**
     * Creates a JID from a string. The format must be
     * <blockquote><p>[ localpart "@" ] domainpart [ "/" resourcepart ]</p></blockquote>
     *
     * @param jid        The JID.
     * @param doUnescape If the jid parameter will be unescaped.
     * @return The JID.
     * @see <a href="http://xmpp.org/extensions/xep-0106.html">XEP-0106: JID Escaping</a>
     */
    public static Jid valueOf(String jid, boolean doUnescape) {
        if (jid == null) {
            throw new IllegalArgumentException("jid must not be null.");
        }
        jid = jid.trim();

        if (jid.isEmpty()) {
            throw new IllegalArgumentException("jid must not be empty.");
        }

        Matcher matcher = JID.matcher(jid);
        if (matcher.matches()) {
            return new Jid(matcher.group(2), matcher.group(3), matcher.group(8), doUnescape, true);
        } else {
            throw new IllegalArgumentException("Could not parse JID.");
        }
    }

    /**
     * Escapes a JID. The characters {@code "&'/:<>@} (+ whitespace) are replaced with their respective escape characters.
     *
     * @param jid The JID.
     * @return The escaped JID.
     * @see <a href="http://xmpp.org/extensions/xep-0106.html">XEP-0106: JID Escaping</a>
     */
    private static String escape(String jid) {
        if (jid != null) {
            Matcher matcher = ESCAPE_PATTERN.matcher(jid);
            StringBuffer sb = new StringBuffer();
            while (matcher.find()) {
                String match = matcher.group();
                matcher.appendReplacement(sb, String.format("\\\\%x", match.getBytes()[0]));
            }
            matcher.appendTail(sb);
            return sb.toString();
        }
        return null;
    }

    private static String unescape(String jid) {
        if (jid != null) {
            Matcher matcher = UNESCAPE_PATTERN.matcher(jid);
            StringBuffer sb = new StringBuffer();
            while (matcher.find()) {
                String match = matcher.group(1);
                int num = Integer.parseInt(match, 16);
                String value = String.valueOf((char) num);
                if (value.equals("\\")) {
                    matcher.appendReplacement(sb, "\\\\");
                } else {
                    matcher.appendReplacement(sb, value);
                }
            }
            matcher.appendTail(sb);
            return sb.toString();
        }
        return null;
    }

    static String prepare(String input, boolean caseFold) {
        if (input != null) {
            String mappedToNothing = MAP_TO_NOTHING.matcher(input).replaceAll("");
            String normalized = Normalizer.normalize(mappedToNothing, Normalizer.Form.NFKC);
            if (caseFold) {
                normalized = normalized.toUpperCase(Locale.ENGLISH).toLowerCase(Locale.ENGLISH);
            }
            // A.7. Notes
            // Because the additional characters prohibited by Nodeprep are
            // prohibited after normalization
            Matcher matcher = PROHIBITED_CHARACTERS.matcher(normalized);
            if (matcher.find()) {
                throw new IllegalArgumentException("Local or resource part contains prohibited characters.");
            }
            return normalized;
        }
        return null;
    }

    private void validateDomain(String domain) {
        if (domain == null) {
            throw new IllegalArgumentException("domain must not be null.");
        }
        if (domain.contains("@")) {
            // Prevent misuse of API.
            throw new IllegalArgumentException("domain must not contain a '@' sign");
        }
        validateLength(domain, "domain");
    }

    /**
     * Validates that the length of a local, domain or resource part is not longer than 1023 characters.
     *
     * @param value The value.
     * @param part  The part, only used to produce an exception message.
     */
    private void validateLength(String value, String part) {
        if (value != null) {
            if (value.isEmpty()) {
                throw new IllegalArgumentException(String.format("%s must not be empty.", part));
            }
            if (value.length() > 1023) {
                throw new IllegalArgumentException(String.format("%s must not be greater than 1023 characters.", part));
            }
        }
    }

    /**
     * Checks if the JID is a full JID.
     * <blockquote>
     * <p>The term "full JID" refers to an XMPP address of the form &lt;localpart@domainpart/resourcepart&gt; (for a particular authorized client or device associated with an account) or of the form &lt;domainpart/resourcepart&gt; (for a particular resource or script associated with a server).</p>
     * </blockquote>
     *
     * @return True, if the JID is a full JID; otherwise false.
     */
    public boolean isFullJid() {
        return resource != null;
    }

    /**
     * Checks if the JID is a bare JID.
     * <blockquote>
     * <p>The term "bare JID" refers to an XMPP address of the form &lt;localpart@domainpart&gt; (for an account at a server) or of the form &lt;domainpart&gt; (for a server).</p>
     * </blockquote>
     *
     * @return True, if the JID is a bare JID; otherwise false.
     */
    public boolean isBareJid() {
        return resource == null;
    }

    /**
     * Converts this JID into a bare JID, i.e. removes the resource.
     * <blockquote>
     * <p>The term "bare JID" refers to an XMPP address of the form &lt;localpart@domainpart&gt; (for an account at a server) or of the form &lt;domainpart&gt; (for a server).</p>
     * </blockquote>
     *
     * @return The bare JID.
     * @see #withResource(String)
     */
    public Jid asBareJid() {
        return new Jid(local, domain, null, false, false);
    }

    /**
     * Returns a new full JID with a resource and the same local and domain part of the current JID.
     *
     * @param resource The resource.
     * @return The full JID with a resource.
     * @see #asBareJid()
     */
    public Jid withResource(String resource) {
        return new Jid(local, domain, resource, false, false);
    }

    /**
     * Gets the local part of the JID, also known as the name or node.
     * <blockquote>
     * <p><cite><a href="http://xmpp.org/rfcs/rfc6122.html#addressing-localpart">2.3.  Localpart</a></cite></p>
     * <p>The localpart of a JID is an optional identifier placed before the domainpart and separated from the latter by the '@' character. Typically a localpart uniquely identifies the entity requesting and using network access provided by a server (i.e., a local account), although it can also represent other kinds of entities (e.g., a chat room associated with a multi-user chat service). The entity represented by an XMPP localpart is addressed within the context of a specific domain (i.e., {@code <localpart@domainpart>}).</p>
     * </blockquote>
     *
     * @return The local part.
     */
    public String getLocal() {
        return local;
    }

    /**
     * Gets the domain part.
     * <blockquote>
     * <p><cite><a href="http://xmpp.org/rfcs/rfc6122.html#addressing-domain">2.2.  Domainpart</a></cite></p>
     * <p>The domainpart of a JID is that portion after the '@' character (if any) and before the '/' character (if any); it is the primary identifier and is the only REQUIRED element of a JID (a mere domainpart is a valid JID).</p>
     * </blockquote>
     *
     * @return The domain part.
     */
    public String getDomain() {
        return domain;
    }

    /**
     * Gets the resource part.
     * <blockquote>
     * <p><cite><a href="http://xmpp.org/rfcs/rfc6122.html#addressing-resource">2.4.  Resourcepart</a></cite></p>
     * <p>The resourcepart of a JID is an optional identifier placed after the domainpart and separated from the latter by the '/' character. A resourcepart can modify either a {@code <localpart@domainpart>} address or a mere {@code <domainpart>} address. Typically a resourcepart uniquely identifies a specific connection (e.g., a device or location) or object (e.g., an occupant in a multi-user chat room) belonging to the entity associated with an XMPP localpart at a domain (i.e., {@code <localpart@domainpart/resourcepart>}).</p>
     * </blockquote>
     *
     * @return The resource part.
     */
    public String getResource() {
        return resource;
    }

    /**
     * Gets the JID in escaped form as described in <a href="http://xmpp.org/extensions/xep-0106.html">XEP-0106: JID Escaping</a>.
     *
     * @return The escaped JID.
     * @see #toString()
     */
    public String toEscapedString() {
        return toString(escapedLocal, domain, resource);
    }

    /**
     * Converts the JID into its string representation, i.e. [ localpart "@" ] domainpart [ "/" resourcepart ].
     *
     * @return The JID.
     * @see #toEscapedString()
     */
    @Override
    public String toString() {
        return toString(local, domain, resource);
    }

    private String toString(String local, String domain, String resource) {
        StringBuilder sb = new StringBuilder();
        if (local != null) {
            sb.append(local);
            sb.append("@");
        }
        sb.append(domain);
        if (resource != null) {
            sb.append("/");
            sb.append(resource);
        }
        return sb.toString();
    }

    @Override
    public boolean equals(Object o) {
        if (o == this) {
            return true;
        }
        if (!(o instanceof Jid)) {
            return false;
        }
        Jid other = (Jid) o;

        return (local == null ? other.local == null : local.equals(other.local))
                && (domain == null ? other.domain == null : domain.equals(other.domain))
                && (resource == null ? other.resource == null : resource.equals(other.resource));

    }

    @Override
    public int hashCode() {
        int result = 17;
        result = 31 * result + ((local == null) ? 0 : local.hashCode());
        result = 31 * result + ((domain == null) ? 0 : domain.hashCode());
        result = 31 * result + ((resource == null) ? 0 : resource.hashCode());
        return result;
    }

    /**
     * Compares this JID with another JID. First domain parts are compared. If these are equal, local parts are compared
     * and if these are equal, too, resource parts are compared.
     *
     * @param o The other JID.
     * @return The comparison result.
     */
    @Override
    public int compareTo(Jid o) {
        if (this == o) {
            return 0;
        }

        if (o != null) {
            int result;
            // First compare domain parts.
            if (domain != null) {
                if (o.domain != null) {
                    result = domain.compareTo(o.domain);
                } else {
                    result = -1;
                }
            } else {
                if (o.domain != null) {
                    result = 1;
                } else {
                    result = 0;
                }
            }
            // If the domains are equal, compare local parts.
            if (result == 0) {
                if (local != null) {
                    if (o.local != null) {
                        result = local.compareTo(o.local);
                    } else {
                        // If this local part is not null, but the other is null, move this down (1).
                        result = 1;
                    }
                } else {
                    // If this local part is null, but the other is not, move this up (-1).
                    if (o.local != null) {
                        result = -1;
                    } else {
                        result = 0;
                    }
                }
            }
            // If the local parts are equal, compare resource parts.
            if (result == 0) {
                if (resource != null) {
                    if (o.resource != null) {
                        result = resource.compareTo(o.resource);
                    } else {
                        // If this resource part is not null, but the other is null, move this down (1).
                        result = 1;
                    }
                } else {
                    // If this resource part is null, but the other is not, move this up (-1).
                    if (o.resource != null) {
                        result = -1;
                    } else {
                        result = 0;
                    }
                }
            }
            return result;
        } else {
            return -1;
        }
    }
}
